File: Exercise_3.cpp
/* 
    Time Complexity                              :  O(2*N) - first 2 reverse operations reverses the array part wise and the last reverse operation reverses the array completely.
    Space Complexity                             :  O(1)
    Did this code successfully run on Leetcode   :  Yes
    Any problem you faced while coding this      :  No
*/

#include <bits/stdc++.h> 
using namespace std; 


class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        useReverse(nums,k);
    }
    
    void useReverse(vector<int>& nums, int k) {
        int n = nums.size();
        k = k%n;
        reverse(begin(nums) + n-k, end(nums));
        reverse(begin(nums), begin(nums)+n-k);
        reverse(begin(nums), end(nums));
    }
};

File: Exercise_3.java
class Solution {
    public void rotate(int[] nums, int k) {
        int n = nums.length;
        int v =0, i =0;
        while(v!=n){
            int p = nums[i], j = (i+k)%n;
            while(j!=i){
                ++v;
                int t = nums[j];
                nums[j] = p;
                p = t;
                j = (j+k)%n;
            }
            ++v;
            nums[i] = p;
            ++i;
        }
    }
}
//tc = O(n)
//sc = O(1)
